javascript/index.d.ts

```diff
@@ -0,0 +1,75 @@
+export type Diff = {
+    0: number;
+    1: string;
+    length: 2;
+    toString: () => string;
+}
+
+export type patch_obj = {
+    diffs: Diff[];
+    start1: number | null;
+    start2: number | null;
+    length1: number;
+    length2: number;
+}
+
+export declare class diff_match_patch {
+    Diff_Timeout: number;
+    Diff_EditCost: number;
+    Match_Threshold: number;
+    Match_Distance: number;
+    Patch_DeleteThreshold: number;
+    Patch_Margin: number;
+    Match_MaxBits: number;
+
+    diff_main(text1: string, text2: string, opt_checklines?: boolean, opt_deadline?: number): Diff[];
+    diff_commonPrefix(text1: string, text2: string): number;
+    diff_commonSuffix(text1: string, text2: string): number;
+    diff_cleanupSemantic(diffs: Diff[]): void;
+    diff_cleanupSemanticLossless(diffs: Diff[]): void;
+    diff_cleanupEfficiency(diffs: Diff[]): void;
+    diff_cleanupMerge(diffs: Diff[]): void;
+    diff_cleanupSplitSurrogates(diffs: Diff[]): void;
+    diff_xIndex(diffs: Diff[], loc: number): number;
+    diff_prettyHtml(diffs: Diff[]): string;
+    diff_text1(diffs: Diff[]): string;
+    diff_text2(diffs: Diff[]): string;
+    diff_levenshtein(diffs: Diff[]): number;
+    diff_toDelta(diffs: Diff[]): string;
+    diff_fromDelta(text1: string, delta: string): Diff[];
+
+    isHighSurrogate(character: string): boolean;
+    isLowSurrogate(character: string): boolean;
+    digit16(character: string): number;
+    decodeURI(text: string): string;
+
+    match_main(text: string, pattern: string, loc: number): number;
+
+    patch_make(a: string, opt_b: string | Diff[]): Array<patch_obj>;
+    patch_make(a: Diff[]): Array<patch_obj>;
+    /** @deprecated use patch_make(a: string, opt_b: Diff[]) */
+    patch_make(a: string, opt_b: string, opt_c: Diff[]): Array<patch_obj>;
```

Comment:
nice work separating these. it's a bit hairy in the code

Reply:
I didn't figure it out from the code - I interpreted the JSDoc comment 🙃 
https://github.com/dmsnell/diff-match-patch/blob/3181e7a79b61117f6862dfcd8c54f294b25c9508/javascript/index.js#L1853-L1859